mmap: Add file descriptor as parameter to PATH

samuel-hunter opened the pull request:
A programmer may wish to create a shared memory object via the POSIX shm API, mmap a descriptor of an anonyous file through some local address socket, or otherwise mmap a file descriptor whose pathname cannot or should not be determined.

This commit enables file descriptors to be passed through the PATH parameter of MMAP on Unix systems. Since it did not open the fd, MUNMAP does not close the fd either.

Although fstat *could* be used to automatically determine the size of the file descriptor, this pass goes for the simpler option and requires the SIZE parameter.

Example use-case:

```lisp
(defvar fd (shm:open-temporary-shm "/blablabla"))
(shm:ftruncate fd 10)

(mmap:with-mmap (addr %fd size fd :size 10 :open '(:read :write))
  (dotimes (i size) 
    (setf (cffi:mem-ref addr :char i) (* i 10)))
  (dotimes (i size)
    (print (cffi:mem-ref addr :char i))))

(shm:close fd)
```

samuel-hunter committed:
mmap: Add file descriptor as parameter to PATH

A programmer may wish to create a shared memory object via the POSIX shm
API, mmap a descriptor of an anonyous file through some local address
socket, or otherwise mmap a file descriptor whose pathname cannot or
should not be determined.

This commit enables file descriptors to be passed through the PATH
parameter of MMAP on Unix systems. Since it did not open the fd, MUNMAP
does not close the fd either.

Although fstat *could* be used to automatically determine the size of
the file descriptor, this pass goes for the simpler option and requires
the SIZE parameter.

samuel-hunter commented on posix.lisp:
```diff
@@ -107,27 +107,35 @@
+        (check-type size unsigned-byte))
       (string
-       (setf fd (u-open path open))
+       (setf fd (u-open path open)
+             fresh-fd fd
+             error-handler (lambda (e)
+                            (declare (ignore e))
+                            (check-posix (= 0 (u-close fd)))))
```
Please double-check my logic here.

I believe the original handler used `(/= fd -1)` as a test to see whether the pathname was null, therefore the fd was generated in the string typecase. Since that's no longer true, I altered the logic so that this handler is used only when the pathname is a string.

Shinmera commented on documentation.lisp:
```diff
@@ -147,6 +147,9 @@ same goes for calling MUNMAP with values not directly returned by MMAP,
 dereference the PTR after a call to MUNMAP.
 
+If an open file descriptor was passed to MMAP, then MUNMAP will not attempt
+to close the file.
```
As discussed in chat, please do not implement it like this. Instead note in MUNMAP that you may pass NIL to forego closure of the FD, and always return a valid FD from MMAP.

samuel-hunter replied:
Ahh right, I forgot about that. 🤦‍♂️

That said (and I think I'm missing something here), I'm not sure how to preserve the fd return value without also closing the file or in some form communicating that the fd ought to be left open.

As you noted, passing in NIL forgoes closure of the fd. This in combination with fd being an opaque value, no other functions using fd, and the caller clearly knowing the fd to be able to pass it in, presents to me that this is the most ergonomic way to do this.

Are you looking for the logic for close-skipping to live in a higher level function, like with-mmap?

Shinmera replied:
My argument is this: if I pass in an FD, I would expect to get the FD back out. It makes no sense to get a "fresh FD" even if it is an opaque value. I may also want to close the FD along with unmapping, so specifying that behaviour in MUNMAP seems reasonable.

As for WITH-MMAP, we could add a keyword argument DONT-CLOSE which will pass NIL to MUNMAP in the unwind.

samuel-hunter replied:
I took another look at WITH-MMAP. Since its paramlist ends with a &REST, I don't believe there's backwards-compatible room for a keyword to indicate not to close.

I *could* add a check to nil-punt the fd if PATH is an int, but that would just bring the same surprising-behavior issue here. Another solution could be to add a WITH-MMAP* to either opt not to close, or to make space for keyword options, though, *euugh*, I really don't like the idea of exposing a second macro doing (almost) the same thing. What are your thoughts?

Shinmera replied:
There is a backwards compatible way. Just add a `&key dont-close` and `(remf args :dont-close)` in the body before passing on the args. The key so far was not used in any backends and as such is completely fine to add.

samuel-hunter replied:
Fixed, thanks :+1:

samuel-hunter committed:
Replace fd NIL-punting in MMAP with DONT-CLOSE flag in WITH-MMAP

MMAP behavior is reverted to always return a valid FD.

Instead of NIL-punting through MMAP to elide closing the FD, this commit
adds the flag DONT-CLOSE to WITH-MMAP. When set, WITH-MMAP replaces the
fd arg in MUNMAP with NIL.

Shinmera committed:
Clarify documentation

Shinmera committed:
Fixnum may be too small.

Shinmera committed:
You must not modify macro arguments

Shinmera commented:
I pushed some more fixes.
